lib/utils/test/src/test_algorithms.cc

```diff
@@ -1,109 +1,128 @@
 #include "doctest.h"
-#include "utils/containers.h"
 #include "utils/graph/adjacency_digraph.h"
 #include "utils/graph/adjacency_multidigraph.h"
 #include "utils/graph/algorithms.h"
 #include "utils/graph/construction.h"
+#include "utils/containers.h"
+#include <iterator>
 
-using namespace FlexFlow::utils;
+using namespace FlexFlow;
 
 TEST_CASE("MultiDiGraph") {
   AdjacencyMultiDiGraph g;
+  Node n0 = g.add_node();
   Node n1 = g.add_node();
   Node n2 = g.add_node();
   Node n3 = g.add_node();
-  Node n4 = g.add_node();
-  MultiDiEdge e1{n1, n4, 0, 0};
-  MultiDiEdge e2{n1, n2, 0, 1};
-  MultiDiEdge e3{n1, n3, 0, 0};
-  MultiDiEdge e4{n2, n3, 0, 0};
+  NodePort p0 = g.add_node_port();
+  NodePort p1 = g.add_node_port();
+  NodePort p2 = g.add_node_port();
+  NodePort p3 = g.add_node_port(); 
+  MultiDiEdge e0{n0, n3, p0, p3};
+  MultiDiEdge e1{n1, n2, p0, p2};
+  MultiDiEdge e2{n1, n3, p1, p3};
+  MultiDiEdge e3{n2, n3, p2, p3};
+  g.add_edge(e0);
   g.add_edge(e1);
   g.add_edge(e2);
   g.add_edge(e3);
-  g.add_edge(e4);
-
-  CHECK(get_nodes(g) == std::unordered_set<Node>{n1, n2, n3, n4});
-  CHECK(get_edges(g) == std::unordered_set<MultiDiEdge>{e1, e2, e3, e4});
-  CHECK(get_incoming_edges(g, {n2, n4}) ==
-        std::unordered_set<MultiDiEdge>{e1, e2});
-  CHECK(get_incoming_edges(g, {n1}) == std::unordered_set<MultiDiEdge>{});
-  CHECK(get_outgoing_edges(g, {n2, n4}) == std::unordered_set<MultiDiEdge>{e4});
-  CHECK(get_predecessors(g, {n1, n2, n3}) ==
-        std::unordered_map<Node, std::unordered_set<Node>>{
+
+  CHECK(g.query_nodes({}) == std::unordered_set<Node>{n0, n1, n2, n3});
+  CHECK(g.query_edges({}) == std::unordered_set<MultiDiEdge>{e0, e1, e2, e3});
+  CHECK(get_incoming_edges(unsafe_create(g), {n1, n3}) ==
+        std::unordered_set<MultiDiEdge>{e0, e2, e3});
+  CHECK(get_incoming_edges(unsafe_create(g), {n1}) == std::unordered_set<MultiDiEdge>{});
```

Comment:
Don't call stuff from algorithms in these tests--should just be tests of the raw interface (yes the previous code did this, but probably shouldn't have)